### ISSUE TITLE
Default to `prod` environment in `.env`?

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/recipes#633]
@danrot said this on Symfony's Slack:

> We have realized that by default the framework-bundle recipe sets the APP_ENV to dev (https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/4.2/manifest.json#L15).
>
> We were wondering, because that means that if somebody is not careful when deploying, the environment will be set to dev on the production system as well, leaking a lot of information.
>
> Wouldn't it be safer to set this to prod by default? Why did you do it the other way round?

And he proposes this nice solution applied in Sulu CMS: https://github.com/sulu/sulu-minimal/pull/189/files

So, maybe we can update this recipe? https://github.com/symfony/recipes/tree/master/symfony/flex/1.0

**Comment by danrot** [comment]
The only drawback I could come up with, is that it only works for `create-project`. If some other team member clones the same project, it will be on `prod` again, because it doesn't create this file anymore. 

But I think then people see the difference, and they can figure out that they have to change the `APP_ENV` variable. I personally would say this tradeoff is ok.

**Comment by Taluu** [comment]
I'm not sure the reverse is not also true. I mean, changing the `APP_ENV` on your prod env is not that much a hassle (it's actually not).

Usually when you need to create a project through the skeleton (and thus flex, so with a recipe on the fw-bundle), it's usually to be directly in the dev-env.

`Sulu` being a CMS, it doesn't apply because it's usually just an app you install (like Wordpress). Not a foundation for an app for you to develop and then deploy.

So :-1: AFAIC.

**Comment by danrot** [comment]
Sulu is not an app you simply install, it also involves some programing before you can get it running for production.

And it is not about being a lot of work to change from `dev` to `prod`, it is about a secure default setting. Beginners might not be aware that they have to do it (although it is not a lot of work), and then they show entire stack traces with a lot of version information in a production environment.

**Comment by javiereguiluz** [comment]
@Taluu what you say is very correct:

```
Usually when you need to create a project through the skeleton
(and thus flex, so with a recipe on the fw-bundle), it's usually
to be directly in the dev-env.
```

But the proposal solves everything: makes `prod` the default in production (`.env` file) and `dev` in your local machine (thanks to the `.env.local` trick).

**Comment by nicolas-grekas** [comment]
Creating a project is not the only moment where things need to be in dev by default. Cloning a project to work on it also is.

Deploying to prod means way more than just setting this variable.

And on proper deployement targets, the real APP_ENV=prod env var should always exist by default.

Not convinced on my side as you spotted :)

**Comment by ndench** [comment]
Doesn't Symfony recommend that you don't deploy the `.env` file anyway? You should be setting environment variables on your servers?

What about other default env vars that get set? For example [doctrine](https://github.com/symfony/recipes/blob/f0e5358bcd7205880fe6894ac6e1e1e02dea4437/doctrine/doctrine-bundle/1.6/manifest.json) sets 

```
DATABASE_URL=mysql://db_user:db_password@127.0.0.1:3306/db_name
```

Which is almost certainly not what you want in production.

**Comment by javiereguiluz** [comment]
@ndench that recommendation changed on November 2018. Now it's recommended to deploy the `.env` file in production. See this: https://symfony.com/doc/current/configuration.html#managing-multiple-env-files

**Comment by fabpot** [comment]
Not convinced either. Let's close.

**Comment by AkashicSeer** [comment]
As a newcomer to Symfony I can concur that environmental variable handling in symfony is  complex and hard to figure out.  In dev mode on the front I dump $_SERVER I can see the values I set in my linux profile script, however they are set to true instead of the actual value WTF is that about.<-- That is what  I don't understand WTF is going on there?

Some of this happened when  I switched from the built in server to nginx using the exact nginx config the docs show. getenv and $_ENV simply quit working, the values were no longer there. I have no idea how the switch broke it.

None of the values I set In the linux script are included in $_ENV anywhere. I am assuming because I have not set the E in GPRSE part of php.ini however if I set a value say DATABASE_URL in my linux profile, then redefine it in env.test so I can access it, symfony decides not to set that value in $_ENV for some odd reason. Which means I have to define the value I need in testing with another name which sucks. I read something in the secrets docs that said not to set an env and Secret to the same name. I found a way to refer to the secret in the yaml file but I can't do that in the .env files because they don't parse the value in env(%%) and return the string instead.

So if I have a secret with name DATABASE_URL I can access it by Autowiring in the services.yaml file and referring to it there. If  I set it in .env.local I can access it $_ENV as long as it doesn't have the same name as one I set in secrets, because then symfony helps me out by not adding it to $_ENV with magic.I honestly would expect anything  I set in env.test to JUST BE INCLUDED instead of doing magic behind the scenes stuff to help protect me from myself.

The problem comes when you want to test something such as a class where you have dependency injected the secrets into the class. Because if you have a class that depends on that class, in order to test it you must mock the values you dependency injected into the class it depends on. 

This means if you have a series of classes that work on a file then upload the file to Amazon S3 for example, and one of those needs the api key, and you define the api key in secrets which sounds smart, you now have no way to access that key in testing.  You can't autowire your classes into phpunit tests,you can't access $container, you can't use $_ENV, you can't use $_SERVER, if you try to redefine the value in .env.local it is ignored....

Your only option after days of failure and digging you finally realize is to create another similar sounding ENV_VARIABLE in .env.test with the key, password, whatever it is then you can use $_ENV['ENV_VARIABLE'] in tests to get the values you need to build the object you need to dependency inject into the class that needs it so you can test it. 

It was a real pain in the ass to figure all of this out. So many lost days because the docs are not clear on this. Shit it took days just to figure out what actual .env files that are still used, so much old and bad info out there. I was told to use env.dev and env.prod on Stackoverflow I don't think those exist anymore according to this https://symfony.com/blog/improvements-to-the-handling-of-env-files-for-all-symfony-versions  

So to recap the only way I have found to access environmental variables inside tests to build dependency injected objects whose params are dependency injected environmental variables is to create alternative names in .env.test and provide the values there, then refer to them with $_ENV['VALUE'] if that helps anyone. you can't use previously defined var names symfony helps you by hiding it if you do.
